# Algoritmo Genetico

In [1]:
import numpy as np

Construimos la funcion objetivo, usa la funcion Rosenbrock.

In [2]:
#Rosenbrock
fob = lambda x: -1*(x[0]**4+3*x[0]**3*x[1]+\
                6*x[0]**2*x[1]**2+3*x[0]*x[1]**3+\
                x[1]**4-4*(x[0]-x[1])*(x[0]+3*x[1]))

Codigo del algoritmo genetivo visto en la diapositivas.

In [3]:
#Generar poblacion
N = 100
P = 2
X0 = 4*np.random.rand(N,P)-2 #Poblacion inicial
X = X0.copy()
#Evaluar poblacion
aptitud = []
for k in range(N):
    aptitud.append(fob(X[k,:]))

aptitud = np.array(aptitud)
indSort = np.argsort(aptitud)

print(max(aptitud), min(aptitud))

for k in range(100):
  #Ruleta
  ValoresRul = aptitud - aptitud.min() + 0.1
  S = sum(ValoresRul)
  r = np.random.rand()*S
  sumasPar = np.cumsum(ValoresRul)

  ind = np.argwhere(sumasPar>=r)[0][0]


  sinInd = list(range(N))
  sinInd.remove(ind)
  S2 = sum(ValoresRul[sinInd])
  r = np.random.rand()*S2
  sumasPar = np.cumsum(ValoresRul[sinInd])

  ind2=np.argwhere(sumasPar>=r)[0][0]


  #cruce
  alpha = np.random.rand()
  hijo = alpha*X[ind,:]+(1-alpha)*X[ind2,:]

  #Mutacion
  if np.random.rand()>0.5:
    cuantos_par = np.random.randint(X.shape[1]+1)
    indHijoMutar = np.random.permutation(X.shape[1])[:cuantos_par]
    hijo[indHijoMutar] += np.random.randn(cuantos_par)

  #Evolucion
  X[indSort[0],:] = hijo
  aptitud[indSort[0]] = fob(hijo)
  indSort = np.argsort(aptitud)

bestInd = np.argmax(aptitud)
mejorX = X[bestInd,:]
mejorAptitud = aptitud[bestInd]
print("La solucion es ",mejorX," con aptitud: ",mejorAptitud)

3.6183330963111584 -140.1932882522647
La solucion es  [1.35405819 0.03211794]  con aptitud:  4.057112105858607


Buscamos en una grilla (grid) el optimo.

In [4]:
x1 = np.linspace(-2,2,100)
x2 = np.linspace(-2,2,100)
x1v, x2v = np.meshgrid(x1,x2)

In [5]:
A = np.hstack((x1v.reshape(100*100,1),x2v.reshape(100*100,1)))

In [6]:
aptitudFB = []
for i in range(A.shape[0]):
  aptitudFB.append(fob(A[i,:]))
aptitudFB = np.asarray(aptitudFB)

In [7]:
np.max(aptitudFB),A[np.argmax(aptitudFB),:]

(4.092306317199881, array([-1.39393939, -0.06060606]))

Segun la grilla, el maximo es 4.09 ubicado en $x_1=-1.39$, $x_2=-0.061$.

# PSO - Enjambre de Particulas

In [8]:
Particulas = X0.copy() #misma poblacion inicial del genetico
F = []
for k in range(N):
    F.append(fob(Particulas[k,:]))

pbest = Particulas.copy()
Fpbest = F.copy()
gbest = Particulas[np.argmax(F),:].copy()
Fgbest = np.max(F)
Iteraciones = 100
c1, c2 = 0.5,0.5
Velocidad = np.zeros_like(Particulas)
for i in range(Iteraciones):
  r1 = np.random.rand(1,P)
  r2 = np.random.rand(1,P)
  #Calculo de la velocidad para todas las particulas
  Velocidad += c1*r1*(pbest-Particulas)+c2*r2*(gbest-Particulas)
  #Actualizacion posicion de todas las particulas
  Particulas += Velocidad
  F = []
  for k in range(N): #Evaluamos fob para cada particula
    F.append(fob(Particulas[k,:]))
    if (F[k]>Fpbest[k]): #Actualizamos el mejor local
      Fpbest[k] = F[k]
      pbest[k,:] = Particulas[k,:].copy()

  maxF = np.max(F)
  if (maxF>Fgbest): #Actualizamos el mejor global
    Fgbest = maxF
    gbest = Particulas[np.argmax(F),:]

print('x: ',gbest,', f(x): ',fob(gbest))

x:  [2.32360074 1.23399266] , f(x):  -114.07708837841487
